### DISCUSSION TITLE
GLib bindings: key names

### DISCUSSION THREAD

**Discussion by mpranj** [ElektraInitiative/libelektra#3667] [Proposals] [answered]
Is there any reason why the GLib bindings make you extract a Key name using properties?
Most other functions are simple getters/setters, so I don't fully understand this decision. Seems inconsistent with the rest of the interface.

https://github.com/ElektraInitiative/libelektra/blob/bb8710053d048ae2742c0a5a578b3fe30b65c1a1/src/bindings/glib/gelektra-key.c#L28-L45

If there is no specific reason, may I at least add a getter for convenience?

**Comment by markus2330** [comment, marked as answer]
Thank you for the question!

Can you make a proposal how the API should look like?

> If there is no specific reason

@manuelm can you maybe answer this?

**Reply by manuelm** [reply]
For API consistence in GI vs. swig. But I don't fully remember anymore. Ofc.

**Reply by mpranj** [reply]
I''d basically just like to conveniently access the key name without modifying it and without having to allocate separate memory for it.
I was thinking of something like:

```C
const gchar * gelektra_key_getname (const GElektraKey * key)
{
	return keyName (key->key);
}
```

In the current form, if I am not mistaken, I have to pass a GValue and then extract the gchar * later which is kind of inconvenient for simply accessing the name.

**Reply by markus2330** [reply]
Makes sense to me to have something like this. About the name: gelektra_key_name is shorter and maybe better?

Or we make a `Key * gelektra_key` then you also have access to unescaped names and so on. `keyName(gelektra_key(k))` is also not terribly long.

**Reply by mpranj** [reply]
> keyName(gelektra_key(k)) is also not terribly long.

Possible, but then we are using the C API again which was not the point of the GLib bindings.

I'm just confused that in the GLib bindings the way to access the name and the value are two completely different approaches and that none of them simply give me a `const gchar *`.

**Reply by manuelm** [reply]
As far as I recall it's because of one or more of:
* GObject Introspection requirement and implementing this way the member access works out of the box
* You should not have access to the underlying data directly
* Some best practice document from glib

**Reply by mpranj** [reply]
> You should not have access to the underlying data directly

This contradicts the implementation of `gelektra_key_getvalue`:
https://github.com/ElektraInitiative/libelektra/blob/bb8710053d048ae2742c0a5a578b3fe30b65c1a1/src/bindings/glib/gelektra-key.c#L360-L363

I do not understand why there are three different ways to access the:
1. name,
2. string value and
3. binary value

of a key.

It is inconsistent in itself. I can add a convenience layer on my side where I use it, but copying the data out also comes at the cost of performance.

**Reply by manuelm** [reply]
It's inconsistent if you look at this as a glib binding. But it was *solely* written to make GI bindings work. Glib bindings alone had no value for me. So *every* line in this file was written so bindings generate nicely.

If you want to understand why things were written that way, move git history back, try changing a line/a return value and see if the GI lua + python bindings still generate.

That said. You are free to change whatever you like without asking me as I don't use it anymore.

PS: There were more properties back in the days: `basename`, `dirname`, `fullname`, etc.

**Reply by mpranj** [reply]
Thank you for the answers!

> It's inconsistent if you look at this as a glib binding. But it was solely written to make GI bindings work. 

Sorry, this was the crucial part I was missing. I _only_ looked at it as glib bindings.

> You are free to change whatever you like without asking me as I don't use it anymore.

I didn't want to change an API without understanding why it was designed as is. 

Since the GI bindings were deprecated and removed, I will just add what I need to the API.
The goal is to make the GSettings bindings work, and it looks like they are actually the only remaining user of the GLib bindings.

**Reply by markus2330** [reply]
> Since the GI bindings were deprecated and removed, I will just add what I need to the API.
> The goal is to make the GSettings bindings work, and it looks like they are actually the only remaining user of the GLib bindings.

Exactly, feel free to do as you see fit. The SWIG bindings are superior to the GI bindings. So if GI bindings do not generate anymore, please remove them.

**Reply by mpranj** [reply]
@markus2330 you removed the GI bindings more than a year ago 039b55b2f66207590e420636faa960807b7b4cdf.
